src/ReverseProxy/Delegation/HttpSysDelegationMiddleware.cs

```diff
@@ -0,0 +1,141 @@
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        _delegationRuleManager = delegationRuleManager ?? throw new ArgumentNullException(nameof(delegationRuleManager));
+        _randomFactory = randomFactory ?? throw new ArgumentNullException(nameof(randomFactory));
+    }
+
+    public Task Invoke(HttpContext context)
+    {
+        _ = context ?? throw new ArgumentNullException(nameof(context));
+        var reverseProxyFeature = context.GetReverseProxyFeature();
+        var destinations = reverseProxyFeature.AvailableDestinations
+            ?? throw new InvalidOperationException($"The {nameof(IReverseProxyFeature)} Destinations collection was not set.");
+        var cluster = reverseProxyFeature.Cluster
+            ?? throw new InvalidOperationException($"The {nameof(IReverseProxyFeature)} Cluster was not set.");
+
+        if (destinations.Any())
+        {
+            // This logic mimics behavior in ForwarderMiddleware, except we save the chosen destination back
+            // to the proxy feature to ensure a delegation destination doesn't slip past this middleware.
+            var destination = destinations[0];
+            if (destinations.Count > 1)
+            {
+                var random = _randomFactory.CreateRandomInstance();
+                Log.MultipleDestinationsAvailable(_logger, reverseProxyFeature.Cluster.Config.ClusterId);
+                destination = destinations[random.Next(destinations.Count)];
+                reverseProxyFeature.AvailableDestinations = destination;
+            }
+
+            if (destination.ShouldUseHttpSysDelegation())
+            {
+                reverseProxyFeature.ProxiedDestination = destination;
+
+                var delegationFeature = context.Features.Get<IHttpSysRequestDelegationFeature>()
+                    ?? throw new InvalidOperationException($"{typeof(IHttpSysRequestDelegationFeature).FullName} is missing.");
+
+                if (!delegationFeature.CanDelegate)
+                {
+                    throw new InvalidOperationException(
+                        "Current request can't be delegated. Either the request body has started to be read or the response has started to be sent.");
+                }
+
+                if (!_delegationRuleManager.TryGetDelegationRule(destination, out var delegationRule))
+                {
+                    Log.DelegationRuleNotFound(_logger, cluster.Config.ClusterId, destination.DestinationId);
+                    context.Response.StatusCode = StatusCodes.Status503ServiceUnavailable;
+                    context.Features.Set<IForwarderErrorFeature>(new ForwarderErrorFeature(ForwarderError.NoAvailableDestinations, ex: null));
+                    return Task.CompletedTask;
+                }
+
+                try
+                {
+                    Log.DelegatingRequest(_logger, delegationRule?.QueueName);
+                    delegationFeature.DelegateRequest(delegationRule!);
```

Comment:
Nullability checks shouldn't be needed here, TryGetDelegationRule shouldn't return null.
```suggestion
                    Log.DelegatingRequest(_logger, delegationRule.QueueName);
                    delegationFeature.DelegateRequest(delegationRule);
```

Reply:
You are correct except for UTs they are needed because I can't create a DelegationRule since the constructor in internal 😞 Any thoughts on a better way to work around this?

Reply:
I thought NotNullWhen would have fixed this.